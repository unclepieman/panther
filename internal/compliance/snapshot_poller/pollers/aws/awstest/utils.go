package awstest

/**
 * Panther is a Cloud-Native SIEM for the Modern Security Team.
 * Copyright (C) 2020 Panther Labs Inc
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import (
	"time"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/arn"
)

// Example output from the AWS API
var (
	ExampleTime, _             = time.Parse(time.RFC3339, "2019-04-02T17:16:30+00:00")
	ExampleIntegrationID       = aws.String("8e39aa9d-9823-4872-a1bd-40fd8795634b")
	ExampleAuthSource          = "arn:aws:iam::123456789012:role/PantherAuditRole"
	ExampleAuthSourceParsedARN = ParseExampleAuthSourceARN(ExampleAuthSource)
	ExampleAccountId           = aws.String("123456789012")
	ExampleRegion              = aws.String("us-west-2")
)

// ParseExampleAuthSourceARN returns a parsed Auth Source ARN
func ParseExampleAuthSourceARN(arnToParse string) arn.ARN {
	parsedArn, err := arn.Parse(arnToParse)
	if err != nil {
		return arn.ARN{}
	}

	return parsedArn
}
