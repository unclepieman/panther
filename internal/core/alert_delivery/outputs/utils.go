package outputs

/**
 * Panther is a Cloud-Native SIEM for the Modern Security Team.
 * Copyright (C) 2020 Panther Labs Inc
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
import (
	"regexp"

	"github.com/aws/aws-sdk-go/aws/awserr"
	"github.com/aws/aws-sdk-go/service/sns"
	"github.com/aws/aws-sdk-go/service/sqs"
	"github.com/pkg/errors"
)

func getAlertResponseFromSQSError(err error) *AlertDeliveryResponse {
	var awsErr awserr.Error
	if errors.As(err, &awsErr) {
		statusCode := mapSQSSendMessageErrorCodeToStatusCode(awsErr)
		return getResponse(statusCode, awsErr.Error())
	}
	return getResponse(500, err.Error())
}

func getAlertResponseFromSNSError(err error) *AlertDeliveryResponse {
	var awsErr awserr.Error
	if errors.As(err, &awsErr) {
		statusCode := mapSNSPublishErrorCodeToStatusCode(awsErr)
		return getResponse(statusCode, awsErr.Error())
	}
	return getResponse(500, err.Error())
}

// getResponse - generates a failed response that can be retried
func getResponse(statusCode int, message string) *AlertDeliveryResponse {
	return &AlertDeliveryResponse{
		StatusCode: statusCode,
		Message:    message,
		Permanent:  false,
		Success:    false,
	}
}

// Maps SNS.Publish error codes to response status codes
func mapSNSPublishErrorCodeToStatusCode(awsErr awserr.Error) int {
	switch awsErr.Code() {
	case sns.ErrCodeInvalidParameterException:
		return 400
	case sns.ErrCodeInvalidParameterValueException:
		return 400
	case sns.ErrCodeInternalErrorException:
		return 500
	case sns.ErrCodeNotFoundException:
		return 404
	case sns.ErrCodeEndpointDisabledException:
		return 403
	case sns.ErrCodePlatformApplicationDisabledException:
		return 403
	case sns.ErrCodeAuthorizationErrorException:
		return 401
	case sns.ErrCodeKMSDisabledException:
		return 403
	case sns.ErrCodeKMSInvalidStateException:
		return 409
	case sns.ErrCodeKMSNotFoundException:
		return 404
	case sns.ErrCodeKMSOptInRequired:
		return 400
	case sns.ErrCodeKMSThrottlingException:
		return 429
	case sns.ErrCodeKMSAccessDeniedException:
		return 401
	case sns.ErrCodeInvalidSecurityException:
		return 401
	default:
		return 500
	}
}

// Maps SQS.SendMessage error codes to response status codes
func mapSQSSendMessageErrorCodeToStatusCode(awsErr awserr.Error) int {
	switch awsErr.Code() {
	case sqs.ErrCodeInvalidMessageContents:
		return 400
	case sqs.ErrCodeUnsupportedOperation:
		return 403
	default:
		return 500
	}
}
func GetOpsGenieRegionalEndpoint(serviceRegion string) string {
	switch serviceRegion {
	case OpsgenieServiceRegionUS:
		return "https://api.opsgenie.com/v2/alerts"
	case OpsgenieServiceRegionEU:
		return "https://api.eu.opsgenie.com/v2/alerts"
	default:
		return "https://api.opsgenie.com/v2/alerts"
	}
}

func removeNewLines(input string) string {
	re := regexp.MustCompile(`\n`)
	return re.ReplaceAllString(input, "")
}
