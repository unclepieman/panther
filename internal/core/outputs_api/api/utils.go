package api

/**
 * Panther is a Cloud-Native SIEM for the Modern Security Team.
 * Copyright (C) 2020 Panther Labs Inc
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import (
	"errors"

	"github.com/aws/aws-sdk-go/aws"
	jsoniter "github.com/json-iterator/go"

	deliverymodel "github.com/panther-labs/panther/api/lambda/delivery/models"
	"github.com/panther-labs/panther/api/lambda/outputs/models"
	"github.com/panther-labs/panther/internal/core/alert_delivery/outputs"
	"github.com/panther-labs/panther/internal/core/outputs_api/table"
	"github.com/panther-labs/panther/pkg/genericapi"
)

const redacted = ""

// AlertOutputToItem converts an AlertOutput to an AlertOutputItem
func AlertOutputToItem(input *models.AlertOutput) (*table.AlertOutputItem, error) {
	item := &table.AlertOutputItem{
		CreatedBy:          input.CreatedBy,
		CreationTime:       input.CreationTime,
		DisplayName:        input.DisplayName,
		LastModifiedBy:     input.LastModifiedBy,
		LastModifiedTime:   input.LastModifiedTime,
		OutputID:           input.OutputID,
		OutputType:         input.OutputType,
		DefaultForSeverity: input.DefaultForSeverity,
		AlertTypes:         input.AlertTypes,
	}

	if input.OutputConfig != nil {
		encryptedConfig, err := encryptionKey.EncryptConfig(input.OutputConfig)
		if err != nil {
			return nil, err
		}
		item.EncryptedConfig = encryptedConfig
	}

	return item, nil
}

// ItemToAlertOutput converts an AlertOutputItem to an AlertOutput
func ItemToAlertOutput(input *table.AlertOutputItem) (alertOutput *models.AlertOutput, err error) {
	alertOutput = &models.AlertOutput{
		CreatedBy:          input.CreatedBy,
		CreationTime:       input.CreationTime,
		DisplayName:        input.DisplayName,
		LastModifiedBy:     input.LastModifiedBy,
		LastModifiedTime:   input.LastModifiedTime,
		OutputID:           input.OutputID,
		OutputType:         input.OutputType,
		DefaultForSeverity: input.DefaultForSeverity,
		AlertTypes:         input.AlertTypes,
	}

	// Decrypt the output before returning to the caller
	alertOutput.OutputConfig = &models.OutputConfig{}
	err = encryptionKey.DecryptConfig(input.EncryptedConfig, alertOutput.OutputConfig)
	if err != nil {
		return nil, err
	}

	return alertOutput, nil
}

func redactOutput(outputConfig *models.OutputConfig) {
	if outputConfig.Slack != nil {
		outputConfig.Slack.WebhookURL = redacted
	}
	if outputConfig.PagerDuty != nil {
		outputConfig.PagerDuty.IntegrationKey = redacted
	}
	if outputConfig.Github != nil {
		outputConfig.Github.Token = redacted
	}
	if outputConfig.Jira != nil {
		outputConfig.Jira.APIKey = redacted
	}
	if outputConfig.Opsgenie != nil {
		outputConfig.Opsgenie.APIKey = redacted
	}
	if outputConfig.MsTeams != nil {
		outputConfig.MsTeams.WebhookURL = redacted
	}
	if outputConfig.Asana != nil {
		outputConfig.Asana.PersonalAccessToken = redacted
	}
	if outputConfig.CustomWebhook != nil {
		outputConfig.CustomWebhook.WebhookURL = redacted
	}
}

// TODO: remove this function when proper migrations are in place
// configureOutputFallbacks - function used to ensure backwards compatibility by backfilling previously missing fields
func configureOutputFallbacks(alertOutput *models.AlertOutput) {
	// Backfill an empty list to contain all the supported types.
	if len(alertOutput.AlertTypes) == 0 {
		alertOutput.AlertTypes = []string{deliverymodel.RuleType, deliverymodel.RuleErrorType, deliverymodel.PolicyType}
	}

	// Backfill an empty Opsgenie service region
	if alertOutput.OutputConfig.Opsgenie != nil {
		if alertOutput.OutputConfig.Opsgenie.ServiceRegion == "" {
			alertOutput.OutputConfig.Opsgenie.ServiceRegion = outputs.OpsgenieServiceRegionUS
		}
	}
}

func getOutputType(outputConfig *models.OutputConfig) (*string, error) {
	if outputConfig.Slack != nil {
		return aws.String("slack"), nil
	}
	if outputConfig.PagerDuty != nil {
		return aws.String("pagerduty"), nil
	}
	if outputConfig.Github != nil {
		return aws.String("github"), nil
	}
	if outputConfig.Jira != nil {
		return aws.String("jira"), nil
	}
	if outputConfig.Opsgenie != nil {
		return aws.String("opsgenie"), nil
	}
	if outputConfig.MsTeams != nil {
		return aws.String("msteams"), nil
	}
	if outputConfig.Sns != nil {
		return aws.String("sns"), nil
	}
	if outputConfig.Sqs != nil {
		return aws.String("sqs"), nil
	}
	if outputConfig.Asana != nil {
		return aws.String("asana"), nil
	}
	if outputConfig.CustomWebhook != nil {
		return aws.String("customwebhook"), nil
	}

	return nil, errors.New("no valid output configuration specified for alert output")
}

// mergeConfigs combines an old config with a new config based on the following rules:
// 1. For every value in the new config, use it
// 2. For every value in the old config, keep it if it is not overwritten by the new config
func mergeConfigs(oldConfig, newConfig *models.OutputConfig) (*models.OutputConfig, error) {
	// Convert the old config into bytes so we can merge it with the new config
	oldBytes, err := jsoniter.Marshal(oldConfig)
	if err != nil {
		return nil, &genericapi.InternalError{
			Message: "Unable to extract existing configuration from dynamo",
		}
	}
	// Turn the bytes into a map so we can work with it more easily
	var oldMap map[string]map[string]interface{}
	err = jsoniter.Unmarshal(oldBytes, &oldMap)
	if err != nil {
		return nil, &genericapi.InternalError{
			Message: "Unable to process existing configuration from dynamo",
		}
	}

	// Repeat for the new config
	newBytes, err := jsoniter.Marshal(newConfig)
	if err != nil {
		return nil, &genericapi.InternalError{
			Message: "Unable to extract the new configuration",
		}
	}
	var newMap map[string]map[string]interface{}
	err = jsoniter.Unmarshal(newBytes, &newMap)
	if err != nil {
		return nil, &genericapi.InternalError{
			Message: "Unable to process the new configuration",
		}
	}

	// Overwrite the existing configurations with the new configurations
	for configType, configMap := range newMap {
		for configKey, configValue := range configMap {
			if configValue == "" {
				continue
			}
			oldMap[configType][configKey] = configValue
		}
	}

	// Turn the map back into bytes
	combinedBytes, err := jsoniter.Marshal(oldMap)
	if err != nil {
		return nil, &genericapi.InternalError{
			Message: "Unable to marshal the combined configuration",
		}
	}

	// Turn the bytes back into a struct
	combinedConfig := &models.OutputConfig{}
	err = jsoniter.Unmarshal(combinedBytes, combinedConfig)
	if err != nil {
		return nil, &genericapi.InternalError{
			Message: "Unable to process the combined configuration",
		}
	}

	return combinedConfig, nil
}

func validateConfigByType(config *models.OutputConfig, outputType *string) error {
	switch *outputType {
	case "slack":
		if config.Slack.WebhookURL != "" {
			return nil
		}
	case "pagerduty":
		if config.PagerDuty.IntegrationKey != "" {
			return nil
		}
	case "github":
		if config.Github.RepoName != "" && config.Github.Token != "" {
			return nil
		}
	case "jira":
		// The Type and AssigneeId are apparently optional, although the frontend requires them
		if config.Jira.APIKey != "" && config.Jira.UserName != "" && config.Jira.ProjectKey != "" && config.Jira.OrgDomain != "" {
			return nil
		}
	case "opsgenie":
		if config.Opsgenie.APIKey != "" {
			return nil
		}
	case "msteams":
		if config.MsTeams.WebhookURL != "" {
			return nil
		}
	case "sns":
		if config.Sns.TopicArn != "" {
			return nil
		}
	case "sqs":
		if config.Sqs.QueueURL != "" {
			return nil
		}
	case "asana":
		if len(config.Asana.ProjectGids) != 0 && config.Asana.PersonalAccessToken != "" {
			return nil
		}
	case "customwebhook":
		if config.CustomWebhook.WebhookURL != "" {
			return nil
		}
	}

	return errors.New("invalid output configuration specified for alert output, missing required fields")
}
