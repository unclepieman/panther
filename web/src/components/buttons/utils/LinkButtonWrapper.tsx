/**
 * Panther is a Cloud-Native SIEM for the Modern Security Team.
 * Copyright (C) 2020 Panther Labs Inc
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
import React from 'react';
import { Box, LinkProps } from 'pouncejs';
import { Link as RRLink } from 'react-router-dom';

export type LinkButtonWrapperProps = Pick<LinkProps, 'external' | 'disabled' | 'to'>;

const LinkButton: React.FC<LinkButtonWrapperProps> = ({ disabled, external, to, children }) => {
  let linkProps: LinkProps;
  if (disabled) {
    linkProps = { as: 'span' as React.ElementType };
  } else if (!external) {
    linkProps = { to, as: RRLink };
  } else {
    linkProps = {
      target: '_blank',
      rel: 'noopener noreferrer',
      href: to as string,
      as: 'a' as React.ElementType,
    };
  }
  return (
    <Box
      {...linkProps}
      sx={{
        '& > span': {
          display: 'flex',
          justifyContent: 'center',
          alignItems: 'center',
        },
      }}
    >
      {children}
    </Box>
  );
};
export default LinkButton;
