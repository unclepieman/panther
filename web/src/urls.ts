/**
 * Panther is a Cloud-Native SIEM for the Modern Security Team.
 * Copyright (C) 2020 Panther Labs Inc
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import {
  AlertDetails,
  ComplianceIntegration,
  Destination,
  GlobalPythonModule,
  LogIntegration,
  Policy,
  ResourceDetails,
  Rule,
  CustomLogRecord,
  DataModel,
  AnalysisPack,
} from 'Generated/schema';

// Typical URL encoding, allowing colons (:) to be present in the URL. Colons are safe.
// https://stackoverflow.com/questions/14872629/uriencode-and-colon
const urlEncode = (str: string) => encodeURIComponent(str).replace(/%3A/g, unescape);

const urls = {
  detections: {
    home: () => '/detections/',
    list: () => urls.detections.home(),
    create: () => `${urls.detections.home()}new/`,
  },
  compliance: {
    home: () => '/cloud-security/',
    overview: () => `${urls.compliance.home()}overview/`,
    policies: {
      list: () => `${urls.detections.list()}policies/`,
      create: () => `${urls.detections.create()}?type=policy`,
      details: (id: Policy['id']) => `${urls.compliance.policies.list()}${urlEncode(id)}/`,
      edit: (id: Policy['id']) => `${urls.compliance.policies.details(id)}edit/`,
    },
    resources: {
      list: () => `${urls.compliance.home()}resources/`,
      details: (id: ResourceDetails['id']) => `${urls.compliance.resources.list()}${urlEncode(id)}/`, // prettier-ignore
      edit: (id: ResourceDetails['id']) => `${urls.compliance.resources.details(id)}edit/`,
    },
  },
  logAnalysis: {
    home: () => '/log-analysis/',
    overview: () => `${urls.logAnalysis.home()}overview/`,
    dataModels: {
      list: () => `${urls.logAnalysis.home()}data-models/`,
      create: () => `${urls.logAnalysis.dataModels.list()}new/`,
      details: (id: DataModel['id']) => `${urls.logAnalysis.dataModels.list()}${urlEncode(id)}/`,
      edit: (id: DataModel['id']) => `${urls.logAnalysis.dataModels.details(id)}edit/`,
    },
    rules: {
      list: () => `${urls.detections.list()}rules/`,
      create: () => `${urls.detections.create()}?type=rule`,
      details: (id: Rule['id']) => `${urls.logAnalysis.rules.list()}${urlEncode(id)}/`,
      edit: (id: Rule['id']) => `${urls.logAnalysis.rules.details(id)}edit/`,
    },
    alerts: {
      list: () => `${urls.logAnalysis.home()}alerts/`,
      details: (id: AlertDetails['alertId']) => `${urls.logAnalysis.alerts.list()}${urlEncode(id)}/` // prettier-ignore
    },
    customLogs: {
      list: () => `${urls.logAnalysis.home()}custom-logs/`,
      details: (logType: CustomLogRecord['logType']) =>
        `${urls.logAnalysis.customLogs.list()}${urlEncode(logType)}/`,
      edit: (logType: CustomLogRecord['logType']) =>
        `${urls.logAnalysis.customLogs.details(logType)}/edit/`,
      create: () => `${urls.logAnalysis.customLogs.list()}new/`,
    },
  },
  packs: {
    home: () => '/packs/',
    list: () => urls.packs.home(),
    details: (id: AnalysisPack['id']) => `${urls.packs.home()}${urlEncode(id)}/`,
  },
  integrations: {
    home: () => `/integrations/`,
    logSources: {
      list: () => `${urls.integrations.home()}log-sources/`,
      create: (type?: string) => `${urls.integrations.logSources.list()}new/${type || ''}`,
      edit: (id: LogIntegration['integrationId'], type: string) =>
        `${urls.integrations.logSources.list()}${type}/${id}/edit/`,
    },
    cloudAccounts: {
      list: () => `${urls.integrations.home()}cloud-accounts/`,
      create: () => `${urls.integrations.cloudAccounts.list()}new/`,
      edit: (id: ComplianceIntegration['integrationId']) =>
        `${urls.integrations.cloudAccounts.list()}${id}/edit/`,
    },
    destinations: {
      list: () => `${urls.integrations.home()}destinations/`,
      create: () => `${urls.integrations.destinations.list()}new/`,
      edit: (id: Destination['outputId']) =>
        `${urls.integrations.destinations.list()}${urlEncode(id)}/edit/`,
    },
  },
  settings: {
    home: () => '/settings/',
    general: () => `${urls.settings.home()}general/`,
    globalPythonModules: {
      list: () => `${urls.settings.home()}global-python-modules/`,
      create: () => `${urls.settings.globalPythonModules.list()}new/`,
      edit: (id: GlobalPythonModule['id']) =>
        `${urls.settings.globalPythonModules.list()}${urlEncode(id)}/edit/`,
    },
    bulkUploader: () => `${urls.settings.home()}bulk-uploader/`,
    users: () => `${urls.settings.home()}users/`,
  },
  account: {
    auth: {
      signIn: () => `/sign-in/`,
      forgotPassword: () => `/password-forgot/`,
      resetPassword: () => `/password-reset/`,
    },
    support: () => '/support',
  },
};

export default urls;
