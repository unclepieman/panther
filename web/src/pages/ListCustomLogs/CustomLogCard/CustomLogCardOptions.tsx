/**
 * Panther is a Cloud-Native SIEM for the Modern Security Team.
 * Copyright (C) 2020 Panther Labs Inc
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import React from 'react';
import { Dropdown, DropdownButton, DropdownMenu, DropdownItem, DropdownLink } from 'pouncejs';
import GenericItemCard from 'Components/GenericItemCard';
import { MODALS } from 'Components/utils/Modal';
import useModal from 'Hooks/useModal';
import urls from 'Source/urls';
import { Link as RRLink } from 'react-router-dom';
import { ListCustomLogSchemas } from '../graphql/listCustomLogSchemas.generated';

interface CustomLogCardOptionsProps {
  customLog: ListCustomLogSchemas['listCustomLogs'][0];
}

const CustomLogCardOptions: React.FC<CustomLogCardOptionsProps> = ({ customLog }) => {
  const { showModal } = useModal();

  return (
    <Dropdown>
      <DropdownButton as={GenericItemCard.OptionsButton} />
      <DropdownMenu>
        <DropdownLink as={RRLink} to={urls.logAnalysis.customLogs.edit(customLog.logType)}>
          Edit
        </DropdownLink>
        <DropdownItem
          onSelect={() =>
            showModal({
              modal: MODALS.DELETE_CUSTOM_LOG,
              props: { customLog },
            })
          }
        >
          Delete
        </DropdownItem>
      </DropdownMenu>
    </Dropdown>
  );
};

export default React.memo(CustomLogCardOptions);
