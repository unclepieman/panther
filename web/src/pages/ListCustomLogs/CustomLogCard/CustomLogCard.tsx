/**
 * Panther is a Cloud-Native SIEM for the Modern Security Team.
 * Copyright (C) 2020 Panther Labs Inc
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import React from 'react';
import { Link, Text } from 'pouncejs';
import GenericItemCard from 'Components/GenericItemCard';
import { Link as RRLink } from 'react-router-dom';
import { formatDatetime } from 'Helpers/utils';
import urls from 'Source/urls';
import { ListCustomLogSchemas } from '../graphql/listCustomLogSchemas.generated';
import CustomLogCardOptions from './CustomLogCardOptions';

interface CustomLogCardProps {
  customLog: ListCustomLogSchemas['listCustomLogs'][0];
}

const CustomLogCard: React.FC<CustomLogCardProps> = ({ customLog }) => {
  return (
    <GenericItemCard>
      <GenericItemCard.Body>
        <GenericItemCard.Header>
          <GenericItemCard.Heading>
            <Link
              as={RRLink}
              to={urls.logAnalysis.customLogs.details(customLog.logType)}
              cursor="pointer"
            >
              {customLog.logType}
            </Link>
          </GenericItemCard.Heading>
          {customLog.referenceURL && (
            <GenericItemCard.HeadingValue
              value={
                <Link external href={customLog.referenceURL}>
                  {customLog.referenceURL}
                </Link>
              }
              label="Reference URL"
              labelFirst
              withDivider
            />
          )}
          <GenericItemCard.HeadingValue
            value={formatDatetime(customLog.updatedAt)}
            label="Updated"
            labelFirst
          />
          <CustomLogCardOptions customLog={customLog} />
        </GenericItemCard.Header>
        <Text fontSize="small">{customLog.description}</Text>
      </GenericItemCard.Body>
    </GenericItemCard>
  );
};

export default CustomLogCard;
