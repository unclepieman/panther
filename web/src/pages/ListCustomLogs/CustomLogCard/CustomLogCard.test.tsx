/**
 * Panther is a Cloud-Native SIEM for the Modern Security Team.
 * Copyright (C) 2020 Panther Labs Inc
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import React from 'react';
import { buildCustomLogRecord, fireClickAndMouseEvents, render } from 'test-utils';
import CustomLogCard from 'Pages/ListCustomLogs/CustomLogCard/CustomLogCard';
import { formatDatetime } from 'Helpers/utils';
import urls from 'Source/urls';

describe('CustomLogCard', () => {
  it('matches snapshot', () => {
    const customLog = buildCustomLogRecord();
    const { container } = render(<CustomLogCard customLog={customLog} />);

    expect(container).toMatchSnapshot();
  });

  it('renders the correct information', () => {
    const customLog = buildCustomLogRecord();
    const { container, getByText } = render(<CustomLogCard customLog={customLog} />);

    expect(getByText(customLog.logType)).toBeInTheDocument();
    expect(getByText(customLog.description)).toBeInTheDocument();
    expect(getByText(customLog.referenceURL)).toBeInTheDocument();
    expect(getByText(formatDatetime(customLog.updatedAt))).toBeInTheDocument();
    expect(
      container.querySelector(`a[href="${urls.logAnalysis.customLogs.details(customLog.logType)}"]`)
    ).toBeTruthy();
  });

  it('renders a dropdown with a delete option', () => {
    const customLog = buildCustomLogRecord();
    const { getByText, getByAriaLabel } = render(<CustomLogCard customLog={customLog} />);

    fireClickAndMouseEvents(getByAriaLabel('Toggle Options'));
    expect(getByText('Delete')).toBeInTheDocument();
  });
});
