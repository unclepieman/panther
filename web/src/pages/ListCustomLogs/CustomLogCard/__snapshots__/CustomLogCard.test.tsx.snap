// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomLogCard matches snapshot 1`] = `
.panther-16 {
  background-color: #1d2a3f;
  border-radius: 4px;
  overflow: hidden;
}

.panther-15 {
  background-color: transparent;
  padding: 16px;
}

.panther-14 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  height: 100%;
}

.panther-13 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  width: 100%;
}

.panther-11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  margin-bottom: 8px;
}

.panther-1 {
  font-weight: 500;
  margin-right: auto;
  max-width: 60%;
  word-break: break-word;
}

.panther-0 {
  -webkit-transition: color 0.1s ease-out;
  transition: color 0.1s ease-out;
  font-weight: 500;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #7fd3ff;
  cursor: pointer;
}

.panther-0:hover {
  color: #ace2ff;
}

.panther-0:focus {
  color: #ace2ff;
}

.panther-0:active,
.panther-0[data-active=true] {
  color: #11a5f3;
}

.panther-4 {
  font-size: 0.75rem;
  max-width: 20%;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.panther-2 {
  margin-right: 4px;
  color: #86a3c3;
}

.panther-3 {
  -webkit-transition: color 0.1s ease-out;
  transition: color 0.1s ease-out;
  font-weight: 500;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #7fd3ff;
}

.panther-3:hover {
  color: #ace2ff;
}

.panther-3:focus {
  color: #ace2ff;
}

.panther-3:active,
.panther-3[data-active=true] {
  color: #11a5f3;
}

.panther-5 {
  padding-left: 8px;
  padding-right: 8px;
  font-size: 0.75rem;
  color: #86a3c3;
}

.panther-10 {
  margin-left: 8px;
  margin-top: -4px;
}

.panther-9 {
  outline: none;
  line-height: 1;
  border-radius: 99999px;
  height: 24px;
  width: 24px;
  -webkit-transition: background-color 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: background-color 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  border: 1px solid;
  border-color: transparent;
  background-color: transparent;
  cursor: pointer;
  color: #f6f6f6;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.panther-9:hover {
  background-color: rgba(62,81,107,0.3);
}

.panther-9:focus {
  background-color: rgba(62,81,107,0.3);
}

.panther-9:active,
.panther-9[data-active=true] {
  background-color: #3e516b;
}

.panther-8 {
  display: inline-block;
  vertical-align: sub;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 16px;
  height: 16px;
  color: currentColor;
}

.panther-12 {
  font-size: 0.75rem;
}

<div>
  <header
    id="main-header"
  />
  <section
    class="panther-16"
  >
    <div
      class="panther-15"
    >
      <div
        class="panther-14"
      >
        <div
          class="panther-13"
        >
          <header
            class="panther-11"
          >
            <h4
              class="panther-1"
            >
              <a
                class="panther-0"
                href="/log-analysis/custom-logs/Towels/"
              >
                Towels
              </a>
            </h4>
            <div
              class="panther-4"
            >
              <span
                class="panther-2"
              >
                Reference URL
              </span>
              <a
                class="panther-3"
                href="Savings Account"
                rel="noopener noreferrer"
                target="_blank"
              >
                Savings Account
              </a>
            </div>
            <span
              class="panther-5"
            >
              â€¢
            </span>
            <div
              class="panther-4"
            >
              <span
                class="panther-2"
              >
                Updated
              </span>
              Invalid Date
            </div>
            <div
              class="panther-10"
            >
              <button
                aria-controls="menu--1"
                aria-haspopup="true"
                aria-label="Toggle Options"
                aria-pressed="false"
                class="panther-9"
                data-reach-menu-button=""
                id="menu-button--menu--1"
                type="button"
              >
                <svg
                  class="panther-8"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M6 10C4.9 10 4 10.9 4 12C4 13.1 4.9 14 6 14C7.1 14 8 13.1 8 12C8 10.9 7.1 10 6 10ZM18 10C16.9 10 16 10.9 16 12C16 13.1 16.9 14 18 14C19.1 14 20 13.1 20 12C20 10.9 19.1 10 18 10ZM12 10C10.9 10 10 10.9 10 12C10 13.1 10.9 14 12 14C13.1 14 14 13.1 14 12C14 10.9 13.1 10 12 10Z"
                  />
                </svg>
              </button>
            </div>
          </header>
          <p
            class="panther-12"
          >
            Rustic
          </p>
        </div>
      </div>
    </div>
  </section>
  <footer
    id="footer"
  />
</div>
`;
